### PR TITLE
perf: Pack the state and future of unfolds in the same memory

### PR DESCRIPTION
**Pull request by Marwes**
cc https://github.com/rust-lang/futures-rs/pull/2268#discussion_r532011280

### CODE REVIEW

File: futures-util/src/sink/unfold.rs [hunk 1]
```diff
@@ -55,23 +61,24 @@ where
 
     fn start_send(self: Pin<&mut Self>, item: Item) -> Result<(), Self::Error> {
         let mut this = self.project();
-        debug_assert!(this.future.is_none());
-        let future = (this.function)(this.state.take().unwrap(), item);
-        this.future.set(Some(future));
+        let future = match this.state.as_mut().project_replace(State::Empty) {
+            StateProjOwn::Value(value) => (this.function)(value, item),
+            StateProjOwn::Future(_) | StateProjOwn::Empty => unreachable!()
```
Comment (taiki-e) [inline futures-util/src/sink/unfold.rs:66]:
```suggestion
            StateProjOwn::Future(_) => panic!("poll_ready not called first"),
            StateProjOwn::Empty => unreachable!()
```

nit: `StateProjOwn::Empty` is really unreachable unless closure panics and the user catch it, but `StateProjOwn::Future` can be reached if the user makes a mistake. A different panic message may be useful.

File: futures-util/src/stream/unfold.rs [hunk 2]
```diff
@@ -74,44 +81,46 @@ where
...
 impl<T, F, Fut, Item> FusedStream for Unfold<T, F, Fut>
-    where F: FnMut(T) -> Fut,
-          Fut: Future<Output = Option<(Item, T)>>,
+where
+    F: FnMut(T) -> Fut,
+    Fut: Future<Output = Option<(Item, T)>>,
 {
     fn is_terminated(&self) -> bool {
-        self.state.is_none() && self.fut.is_none()
+        matches!(self.state, State::Empty)
```
Comment (taiki-e) [inline futures-util/src/stream/unfold.rs:94]: `matches!` macro is not available in our MSRV (1.37). Could you replace it with `match` or `if let`?

File: futures-util/src/stream/unfold.rs [hunk 3]
```diff
@@ -74,44 +81,46 @@ where
...
-    fn poll_next(
-        self: Pin<&mut Self>,
-        cx: &mut Context<'_>,
-    ) -> Poll<Option<Self::Item>> {
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         let mut this = self.project();
 
-        if let Some(state) = this.state.take() {
-            this.fut.set(Some((this.f)(state)));
+        if matches!(&*this.state, State::Value(_)) {
```
Comment (taiki-e) [inline futures-util/src/stream/unfold.rs:108]: Same here

**Review by taiki-e — COMMENTED** [review]
Thanks! This looks good to me aside from a few nits.

File: futures-util/src/sink/unfold.rs [hunk 4]
```diff
@@ -55,27 +61,33 @@ where
...
+        Poll::Ready(
+            if let StateProj::Future(future) = this.state.as_mut().project() {
+                let result = match ready!(future.poll(cx)) {
+                    Ok(state) => {
+                        this.state.set(State::Value(state));
+                        Ok(())
+                    }
+                    Err(err) => Err(err),
+                };
+                result
```
Comment (taiki-e) [inline futures-util/src/sink/unfold.rs:86]:
```suggestion
                match ready!(future.poll(cx)) {
                    Ok(state) => {
                        this.state.set(State::Value(state));
                        Ok(())
                    }
                    Err(err) => Err(err),
                }
```

**Comment by Marwes** [comment]
Had to updated this to manual pin projections due to https://github.com/rust-lang/futures-rs/pull/2273 .

**Review by taiki-e — CHANGES_REQUESTED** [review]
Thanks. The implementation LGTM, but seems CI is still failed. Could you fix it? (probably you need to replace `std` with `core` in `unfold_state.rs`)

In addition, there are many irrelevant format changes in `lib.rs`, could you remove them?

File: futures-util/src/unfold_state.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
+use std::pin::Pin;
```
Comment (taiki-e) [inline futures-util/src/unfold_state.rs:1]:
```suggestion
use core::pin::Pin;
```

File: futures-util/src/unfold_state.rs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                _ => None,
+            }
+        }
+    }
+
+    pub(crate) fn take_value(self: Pin<&mut Self>) -> Option<T> {
+        // SAFETY We only move out of the `Value` variant which is not pinned
+        match *self {
+            Self::Value(_) => unsafe {
+                match std::mem::replace(self.get_unchecked_mut(), UnfoldState::Empty) {
```
Comment (taiki-e) [inline futures-util/src/unfold_state.rs:26]:
```suggestion
                match core::mem::replace(self.get_unchecked_mut(), UnfoldState::Empty) {
```

File: futures-util/src/unfold_state.rs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                    UnfoldState::Value(v) => Some(v),
+                    _ => std::hint::unreachable_unchecked(),
```
Comment (taiki-e) [inline futures-util/src/unfold_state.rs:28]:
```suggestion
                    _ => core::hint::unreachable_unchecked(),
```

File: futures-util/src/sink/unfold.rs [hunk 8]
```diff
@@ -1,4 +1,4 @@
-use crate::unfold_state::UnfoldState;
+se crate::unfold_state::UnfoldState;
```
Comment (piegamesde) [inline futures-util/src/sink/unfold.rs:1]: Typo?

**Review by taiki-e — APPROVED** [review]

**Pull request merged**